src/java.base/share/classes/java/text/DecimalFormat.java

```diff
@@ -3712,7 +3712,9 @@ else if (ch == percent) {
             setMinimumIntegerDigits(0);
             setMaximumIntegerDigits(MAXIMUM_INTEGER_DIGITS);
             setMinimumFractionDigits(0);
-            setMaximumFractionDigits(MAXIMUM_FRACTION_DIGITS);
+            // As maxFracDigits are fully displayed unlike maxIntDigits
+            // Prevent OOME by setting to a much more reasonable value.
+            setMaximumFractionDigits(DOUBLE_FRACTION_DIGITS);
```

Comment:
Setting a reasonable default makes sense.  
In other control paths, the max fraction digits come from the inputs or are explicitly set.

It might be a reasonable related change to use StringBuilder.repeat() instead of a loop at LIne 3312-3319, where the pattern char(s) are being appended to the result.

Reply:
Thanks for taking a look. Updated the loop with `repeat` as you suggested, and decided to refactor the rest of the method while I was at it.

Additionally, I added some more tests, as it seems that there is a lack of pattern tests for DecimalFormat in general.

> In other control paths, the max fraction digits come from the inputs or are explicitly set.

Right, while `Integer.MAX_VALUE` can still be set by other control paths (and thus an OOME if toPattern() is invoked), this is something explicitly done by the user, and thus we decided we would still allow the behavior.